### ISSUE TITLE
Thoughts on "HTML Signing Profile"?

### ISSUE THREAD

**Issue by Summertime** [tasn/webext-signed-pages#20]
https://www.w3.org/2007/11/h6n/

I'm not a fan of the entire thing, but, it does seem like it could possibly be safer:

at the moment, there is 3 separate parsers all working on the same file: the minifier parser (to be removed once chrome is fixed though), the browser parser, and the parser to find the signature. If there is a mismatch for any of those, then that's an opportunity for something to potentially be snuck through

for something based on HTMLSP, the process would instead be: the HTML page has a link to a file providing proof of the integrity of the document, the file signing the document explicitly states the URL(s) and their hashes, and then that list of urls and hashes gets signed.

As a result, only 1 parser is doing work (The browser's parser can be used to find the link, without executing javascript or fetching resources via `DOMParser`, [however this might cause other issues](https://www.youtube.com/watch?v=lG7U3fuNw3A)), downsides of this are that it can cause slowdown (probably too much slowdown without http2 / server-push)

Another thought would be to try and embed a signature into the HTTP headers, though there might not be enough space

obviously I wouldn't suggest such suffering as using XMLDSIG as they suggest, if JSON were to be picked instead, making a canonical version of the `SigningInfo` block would be a whole lot easier.

And it would obviously complicate the implementation a lot

**Comment by rugk** [comment]
I guess https://www.w3.org/2007/11/h6n/ would be implemented in browsers, making this add-on obsolete though.

**Comment by tasn** [comment]
@rugk, it doesn't look like it ever will. It's been over a decade. This plugin "solves" adoption by bypassing the browser committees altogether. Let's first get widespread adoption and then make a standard. :)

@Summertime, I agree with your comment regarding the 3 parsers. As you said, the first one will be dropped once Chrome implements the needed API. I'm even considering making it into 2 signatures one with the minifier and one without (for browsers that support this API) so at least Firefox users are safer. It shouldn't be too much work and while I was initially against two code paths, I think the benefits outweigh the costs. Additionally, if you take a look at #18, you'll see that moving the signature to a meta tag is planned (which also involve changing to the browser's parser) which solves your other concern. Making all of the chain to use the save parser.

I'm very much against having any external files due to performance and complexity. We have to block page load when checking the signature, so depending on another network request would be problematic.

I'm against HTTP headers because it means deployment is much more difficult, or in some cases (e.g. Github pages) impossible. At the moment you can just serve a simple HTML file and get signature checks. This is not the case with HTTP headers.


In summary, I don't really have thoughts on the standard, but I'm against external files, using headers and etc. With that being said, many of your concerns are valid, though are already addressed in other tickets/PRs. I'm therefore closing this issue, as I think I answered your questions and I don't think there's anything more to add. Please reopen if you feel like this is not the case.

Thanks!
